File: src/com/themaskedcrusader/tmcz/modules/bleed/BleedSchedule.java
/*
 * Copyright 2013 Topher Donovan (themaskedcrusader.com)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package com.themaskedcrusader.tmcz.modules.bleed;

import com.themaskedcrusader.bukkit.config.Messages;
import com.themaskedcrusader.bukkit.config.Settings;
import com.themaskedcrusader.tmcz.data.PlayerUtil;
import com.themaskedcrusader.tmcz.data.Status;
import org.bukkit.ChatColor;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;

public class BleedSchedule {
    private static JavaPlugin plugin;

    public BleedSchedule(JavaPlugin plugin) {
        BleedSchedule.plugin = plugin;

        plugin.getServer().getScheduler().runTaskTimerAsynchronously(plugin, new Runnable() {
            public void run() {
                BleedSchedule.bleedThePlayer();
            }
        }, 20L, (Settings.getConfig().getLong(Bleed.TICKS)));
    }

    @SuppressWarnings("unchecked")
    public static void bleedThePlayer() {
        Player[]  players = plugin.getServer().getOnlinePlayers();
        for (Player player : players) {
            if (Bleed.isAllowed(player) && PlayerUtil.isStatusSet(player, Status.BLEEDING)) {
                int seconds = Settings.getConfig().getInt(Bleed.SECONDS) * 20;
                int damage = Settings.getConfig().getInt(Bleed.DAMAGE) * 13;
                PotionEffect bleedPotion = new PotionEffect(PotionEffectType.getById(15), seconds, 1);
                PotionEffect bleedDamage = new PotionEffect(PotionEffectType.getById(19), damage, 1);
                player.addPotionEffect(bleedPotion);
                player.addPotionEffect(bleedDamage);
                player.sendMessage(ChatColor.RED + Messages.getConfig().getString(Bleed.MESSAGE));
            }
        }
    }
}

File: src/main/kotlin/me/schalk/spigot/tmcz/modules/bleed/BleedSchedule.kt
/*
 * Copyright (c) 2013 - 2023 Christopher Schalk
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package me.schalk.spigot.tmcz.modules.bleed

import me.schalk.spigot.lib.config.getMessages
import me.schalk.spigot.lib.config.getSettings
import me.schalk.spigot.tmcz.data.GameData
import org.bukkit.ChatColor
import org.bukkit.Particle
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin

class BleedSchedule(plugin: JavaPlugin) : BleedModule() {

    init{
        BleedSchedule.plugin = plugin
        plugin.server.scheduler.runTaskTimerAsynchronously(
            plugin,
            Runnable { bleedThePlayer() }, 20L, getSettings().getConfig().getLong(MODULE + TICKS)
        )
    }

    companion object {
        lateinit var plugin: JavaPlugin

        fun bleedThePlayer() {
            plugin.server.onlinePlayers.forEach{ player ->
                if (isAllowed(player) && GameData.getPlayer(player).bleeding) {
                    plugin.server.scheduler.runTaskLater(plugin, Runnable {
                        player.damage(getSettings().getConfig().getDouble(MODULE + DAMAGE))
                        player.spawnParticle(Particle.DAMAGE_INDICATOR, player.location, getSettings().getConfig().getInt(MODULE + PARTICLES),  0.2, 0.2, 0.2, 0.01)
                                                                          }, 10)
                    player.sendMessage(ChatColor.RED.toString() + getMessages().getConfig().getString(MODULE + MESSAGE))
                }
            }
        }
    }
}